### ISSUE TITLE
Add an interactive shell for Sqlite3

### ISSUE THREAD

**Issue by rhettinger** [python/cpython#77617]
BPO | [33436](https://bugs.python.org/issue33436)
--- | :---
Nosy | @rhettinger, @bitdancer, @serhiy-storchaka, @erlend-aasland, @DiddiLeija
Files | <li>[sqlite3_repl.py](https://bugs.python.org/file47573/sqlite3_repl.py "Uploaded as text/plain at 2018-05-07.01:46:58 by @rhettinger"): Minimal proof-of-concept</li><li>[0001-Add-proof-of-concept-REPL.patch](https://bugs.python.org/file49199/0001-Add-proof-of-concept-REPL.patch "Uploaded as text/plain at 2020-05-28.09:17:47 by @erlend-aasland")</li>

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = None
closed_at = None
created_at = <Date 2018-05-07.01:47:00.123>
labels = ['3.8', '3.9', '3.10', '3.11', 'type-feature', 'library']
title = 'Add an interactive shell for Sqlite3'
updated_at = <Date 2022-02-11.14:45:10.289>
user = 'https://github.com/rhettinger'
```

bugs.python.org fields:
```python
activity = <Date 2022-02-11.14:45:10.289>
actor = 'DiddiLeija'
assignee = 'none'
closed = False
closed_date = None
closer = None
components = ['Library (Lib)']
creation = <Date 2018-05-07.01:47:00.123>
creator = 'rhettinger'
dependencies = []
files = ['47573', '49199']
hgrepos = []
issue_num = 33436
keywords = ['patch']
message_count = 6.0
messages = ['316248', '316257', '316261', '327252', '370180', '413072']
nosy_count = 6.0
nosy_names = ['rhettinger', 'r.david.murray', 'serhiy.storchaka', 'Big Stone', 'erlendaasland', 'DiddiLeija']
pr_nums = []
priority = 'normal'
resolution = None
stage = None
status = 'open'
superseder = None
type = 'enhancement'
url = 'https://bugs.python.org/issue33436'
versions = ['Python 3.8', 'Python 3.9', 'Python 3.10', 'Python 3.11']
```

</p></details>

**Comment by erlend-aasland** [comment]
@rhettinger, @serhiy-storchaka, are you still +1 on this? I'm +1, because...

- implementation is easy; low maintenance cost
- I'm a fan of built-in tools like this (`python -m calendar` is one of my favourites)
- there has been _some_ interest in this feature (re. the linked Discourse topic)

**Comment by serhiy-storchaka** [comment]
I am +1 on this for the same reasons. Go ahead, it would be fun.

**Comment by kumaraditya303** [comment]
I am +1, this would make Windows sqlite users happy.

**Comment by erlend-aasland** [comment]
I'll revive my patch.

**Comment by erlend-aasland** [comment]
> I'll revive my patch.

Rebased and cleaned up slightly. Draft PR up: #95026

IMO, let's just provide a dead simple sqlite3 shell for now; if more features are needed, let's add those in separate issues.

**Comment by erlend-aasland** [comment]
Possible enhancements:

- provide switches for overriding row and text factories
- add more SQLite CLI features:
    - `-cmd COMMAND`: run "COMMAND" before reading stdin
    - `-deserialize`: open the database using `sqlite3_deserialize()`
    - `-bail`: stop after hitting an error
    - `-readonly`: open the database read-only
   - `.trace` command: set trace callback to `print`
   - `.progress` command: adds a basic progress handler
   - `.connection [close] [#]: Open or close an auxiliary database connection

**Comment by erlend-aasland** [comment]
Thanks Raymond, for the pitch. Thanks Dong-hee, Kumar, and Serhiy for helpful reviews. This has now been implemented by #95026. Enhancement requests go as separate issues.
